### ISSUE TITLE
Resolving errors with Linux Mint

### ISSUE THREAD

**Issue by sylvainpelissier** [rickysarraf/apt-offline#59]
I'm trying to update a Linux Mint 18 distribution but I have the current errors on the offline machine:

```
$ sudo apt-get upgrade
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  firefox firefox-locale-en mint-mirrors mint-upgrade-info mintinstall mintupdate ubuntu-drivers-common
7 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 49.4 MB of archives.
After this operation, 2'432 kB of additional disk space will be used.
Do you want to continue? [Y/n] y
Err:1 http://packages.linuxmint.com sarah/upstream amd64 ubuntu-drivers-common amd64 1:0.4.17.2linuxmint1
  Temporary failure resolving 'packages.linuxmint.com'
Err:2 http://packages.linuxmint.com sarah/upstream amd64 firefox amd64 50.1.0+linuxmint1+serena
  Temporary failure resolving 'packages.linuxmint.com'
Err:3 http://packages.linuxmint.com sarah/upstream amd64 firefox-locale-en amd64 50.1.0+linuxmint1+serena
  Temporary failure resolving 'packages.linuxmint.com'
Ign:4 http://packages.linuxmint.com sarah/main i386 mint-mirrors all 1.1.5
Ign:5 http://packages.linuxmint.com sarah/main i386 mint-upgrade-info all 1.0.6
Ign:6 http://packages.linuxmint.com sarah/main i386 mintinstall all 7.7.5
Ign:7 http://packages.linuxmint.com sarah/main i386 mintupdate all 5.1.0.4
Err:4 http://packages.linuxmint.com sarah/main i386 mint-mirrors all 1.1.5
  Temporary failure resolving 'packages.linuxmint.com'
Err:5 http://packages.linuxmint.com sarah/main i386 mint-upgrade-info all 1.0.6
  Temporary failure resolving 'packages.linuxmint.com'
Err:6 http://packages.linuxmint.com sarah/main i386 mintinstall all 7.7.5
  Temporary failure resolving 'packages.linuxmint.com'
Err:7 http://packages.linuxmint.com sarah/main i386 mintupdate all 5.1.0.4
  Temporary failure resolving 'packages.linuxmint.com'
E: Failed to fetch http://packages.linuxmint.com/pool/upstream/u/ubuntu-drivers-common/ubuntu-drivers-common_0.4.17.2linuxmint1_amd64.deb  Temporary failure resolving 'packages.linuxmint.com'

E: Failed to fetch http://packages.linuxmint.com/pool/upstream/f/firefox/firefox_50.1.0+linuxmint1+serena_amd64.deb  Temporary failure resolving 'packages.linuxmint.com'

E: Failed to fetch http://packages.linuxmint.com/pool/upstream/f/firefox/firefox-locale-en_50.1.0+linuxmint1+serena_amd64.deb  Temporary failure resolving 'packages.linuxmint.com'

E: Failed to fetch http://packages.linuxmint.com/pool/main/m/mint-mirrors/mint-mirrors_1.1.5_all.deb  Temporary failure resolving 'packages.linuxmint.com'

E: Failed to fetch http://packages.linuxmint.com/pool/main/m/mint-upgrade-info/mint-upgrade-info_1.0.6_all.deb  Temporary failure resolving 'packages.linuxmint.com'

E: Failed to fetch http://packages.linuxmint.com/pool/main/m/mintinstall/mintinstall_7.7.5_all.deb  Temporary failure resolving 'packages.linuxmint.com'

E: Failed to fetch http://packages.linuxmint.com/pool/main/m/mintupdate/mintupdate_5.1.0.4_all.deb  Temporary failure resolving 'packages.linuxmint.com'

E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

**Comment by rickysarraf** [comment]
As I understand from your issue, you want to upgrade this machine, and this machine doesn't have internet access ?

If so, please follow the howto: https://www.researchut.com/tools/apt-offline

**Comment by sylvainpelissier** [comment]
Yes that's it. I followed the instructions but the last command
```
apt-get upgrade now
```
gives the previous errors. I just notices that before when I run the command:
```
sudo apt-offline install mint-install.zip 
```
I got the following errors:

```
No changelog available
(Y) Yes. Proceed with installation
(N) No, Abort.
(C) Display changelog
(?) Display this help message.
What would you like to do next:	 (y, N, ?)y
Proceeding with installation
ERROR: Failed to sync file /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_binary-amd64_Packages
WARN: Failed to unlink temproary file /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_binary-amd64_Packagesapt-offline. Check respective decompressor library support
ERROR: Failed to sync file /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_binary-i386_Packages
WARN: Failed to unlink temproary file /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_binary-i386_Packagesapt-offline. Check respective decompressor library support
ERROR: Failed to sync file /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_i18n_Translation-en
WARN: Failed to unlink temproary file /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_i18n_Translation-enapt-offline. Check respective decompressor library support
```
...

**Comment by rickysarraf** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

On Fri, 2016-12-23 at 03:28 -0800, Sylvain Pelissier wrote:
> No changelog available
> (Y) Yes. Proceed with installation
> (N) No, Abort.
> (C) Display changelog
> (?) Display this help message.
> What would you like to do next:	 (y, N, ?)y
> Proceeding with installation
> ERROR: Failed to sync file
> /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_b
> inary-amd64_Packages
> WARN: Failed to unlink temproary file
> /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_b
> inary-amd64_Packagesapt-offline. Check respective decompressor library support
> ERROR: Failed to sync file
> /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_b
> inary-i386_Packages
> WARN: Failed to unlink temproary file
> /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_b
> inary-i386_Packagesapt-offline. Check respective decompressor library support
> ERROR: Failed to sync file
> /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_i
> 18n_Translation-en
> WARN: Failed to unlink temproary file
> /var/lib/apt/lists/partial/archive.canonical.com_ubuntu_dists_xenial_partner_i
> 18n_Translation-enapt-offline. Check respective decompressor library support

Yes. This is usually an error that shows up if your disconnected machine does
not already have python-lzma package installed. Can you please check that first
?


- -- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEQCVDstmIVAB/Yn02pjpYo/LhdWkFAlheU7sACgkQpjpYo/Lh
dWm7xA//SL5z7X0LsIZKINpB5Esy7wyX9itJAgDNr/wVvPX4TwCTCVUVJBtw/FpS
36LFMGDhyPP3PorENvNfQZvcgK7Kbl1qsGdsPC8C4oPWFYtoQFrd2HwneCE0Uz3s
PbOeMauWXnDSOzCh6xXhB1Rwgf6i0cvhTxJvgAB4kQ8M6EmDhE6Bn2H0s09CgCWw
mAb7/DJoFMnbTsNPZe0FDtgBR/a73cBS3LF/83Lz6MTILqLEAnxuVo+2HEygXQEG
76JgguRtwC1KNJ+TA5/kXbWb98njoRddZDD8ycCF/K5hIf4CUx+WRvGWyOpyo62c
3u+AiBahNrg0A3pKbwnyR78ge9lEwvDDqeIgeDE9ulXaBvF/iGdiX+ngY3q3PnD9
uhkk+u0QBEQxDVYKvVLqtdynUz8l41o93h3ZKzqx9GpgHuNRdOGXZbwcsxCCmFDF
VDDCB4khT5NT3TuFmDti2DcLItooEN8Pv7kTnlyzHyHXatGS9/4G56hK2KcX032r
WvOETopiNJzzYqk9QTNKcSGd+XMBm2Nd571vNpT8GBylWnWWHA8+LEy0B83oBiB+
pc3ohCSMPpfXouXxDDaI5yb55VcLHkqtjvctBPYJSqza/3yRk2FZeDlnko7/nSrO
Oi1cxUMxxurAVEQBRgyzAwK5Wkt4G6Q3TYBwc7dX3Jw8ty636aE=
=3qaa
-----END PGP SIGNATURE-----

**Comment by sylvainpelissier** [comment]
Thanks for your answers. After installing python-lzma package I have no more error during the command:
```
sudo apt-offline install mint-install.zip 
```

However, after the command:
```
apt-get upgrade
```

I have the following error messages:
```
102 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 181 MB of archives.
After this operation, 35.1 MB of additional disk space will be used.
Do you want to continue? [Y/n] y
Ign:1 http://mirrors.ircam.fr/pub/linuxmint sarah/upstream i386 base-files all 18.0.2
Err:2 http://mirror.switch.ch/ftp/mirror/ubuntu xenial-updates/main amd64 libc6-dbg amd64 2.23-0ubuntu5
  Temporary failure resolving 'mirror.switch.ch'
Err:3 http://mirrors.ircam.fr/pub/linuxmint sarah/upstream amd64 firefox amd64 50.1.0+linuxmint1+serena
  Temporary failure resolving 'mirrors.ircam.fr'
Err:4 http://mirror.switch.ch/ftp/mirror/ubuntu xenial-updates/main amd64 libc6-dev amd64 2.23-0ubuntu5
  Temporary failure resolving 'mirror.switch.ch'
Err:5 http://mirrors.ircam.fr/pub/linuxmint sarah/upstream amd64 firefox-locale-en amd64 50.1.0+linuxmint1+serena
  Temporary failure resolving 'mirrors.ircam.fr'
Err:6 http://mirror.switch.ch/ftp/mirror/ubuntu xenial-updates/main amd64 libc-dev-bin amd64 2.23-0ubuntu5
  Temporary failure resolving 'mirror.switch.ch'
Err:7 http://mirrors.ircam.fr/pub/linuxmint sarah/upstream amd64 libplymouth4 amd64 0.9.2+mint1+sarah
  Temporary failure resolving 'mirrors.ircam.fr'
Err:8 http://security.ubuntu.com/ubuntu xenial-security/main amd64 linux-libc-dev amd64 4.4.0-57.78
  Temporary failure resolving 'mirror.switch.ch'
Ign:9 http://mirrors.ircam.fr/pub/linuxmint sarah/main i386 mint-upgrade-info all 1.0.6
Ign:10 http://mirrors.ircam.fr/pub/linuxmint sarah/main i386 mintupdate all 5.1.0.4
Err:11 http://mirrors.ircam.fr/pub/linuxmint sarah/upstream amd64 plymouth-label amd64 0.9.2+mint1+sarah
  Temporary failure resolving 'mirrors.ircam.fr'
Err:12 http://mirrors.ircam.fr/pub/linuxmint sarah/upstream amd64 plymouth amd64 0.9.2+mint1+sarah
  Temporary failure resolving 'mirrors.ircam.fr'
Err:13 http://mirrors.ircam.fr/pub/linuxmint sarah/upstream amd64 plymouth-theme-ubuntu-text amd64 0.9.2+mint1+sarah
...
```

Did I miss something else ?

**Comment by rickysarraf** [comment]
To make any conclusion on what may have happened, I need the apt-offline install command's console output. And also the apt-offline get command's console output.

apt-offline will do its bets to download all data mentioned in the signature file. But not all data may be available on the remote server. 
Reasons:
1. Old data may have been superseded by newer ones
2. Transport errors will make apt-offline discard that data.

At the end of the apt-offline get operation, it'd tell you what all it failed to download.

In your above output, it is clear something failed. Because out of the 181 MiB, 35.1 MiB is pending.

Also, what all options did you use during the apt-offline set operation ?
You can play around with its "set --upgrade --upgrade-type dist-upgrade" option.

**Comment by sylvainpelissier** [comment]
Here is the apt-offline get command's console output:
https://framabin.org/?0e346d5271fc56ef#k/K9VBHHSrtYQhS/kUumKyZ+BvjfyiXjwU5xg1U0Xa8=

Here is the apt-offline install command's console output (no error):
https://framabin.org/?31aa24aa8fa19154#WFBz3cvaUaxqf6zfbYLnOD3BfRf2PzfP4UEGS6N5K7M=

and here the apt-get upgrade  command's console output:
https://framabin.org/?7773b1e63afdac34#5cnZPTjLLx7DQOL8zkXaPRpY5C2PzMpMONimQPOdt4g=

I used the following command:
```
sudo apt-offline set mint-update.sig --update
```
for the set operation.

**Comment by rickysarraf** [comment]
The set command that you ran is the problem. Quoting the help:

```
rrs@learner:~$ apt-offline set --help
usage: apt-offline set [-h] [--verbose] [--simulate]
                       [--install-packages [PKG [PKG ...]]]
                       [--install-src-packages [SOURCE PKG [SOURCE PKG ...]]]
                       [--src-build-dep] [--release release_name] [--update]
                       [--upgrade] [--upgrade-type upgrade]
                       [--generate-changelog] [--apt-backend apt-get]
                       apt-offline.sig

positional arguments:
  apt-offline.sig       Generate a signature file

optional arguments:
  -h, --help            show this help message and exit
  --verbose             Enable verbose messages
  --simulate            Just simulate. Very helpful when debugging
  --install-packages [PKG [PKG ...]]
                        Packages that need to be installed
  --install-src-packages [SOURCE PKG [SOURCE PKG ...]]
                        Source Packages that need to be installed
  --src-build-dep       Install Build Dependency packages for requested source
                        packages
  --release release_name
                        Release target to install packages from
  --update              Generate Signature to update APT Database
  --upgrade             Generate Signature of packages to be upgraded
  --upgrade-type upgrade
                        Type of upgrade to do. Use one of upgrade, dist-
                        upgrade, dselect-ugprade
  --generate-changelog  Generate changelog of the version to be downloaded
  --apt-backend apt-get
                        APT backend to use. One of: apt, apt-get, python-apt
2017-01-05 / 15:46:34 ♒♒♒  ☺  
```


The --update command will generate the list of data needed to only update your apt database. The equivalent of `apt-get update`

The --upgrade command will generate the list of data needed to upgrade your packages. The equivalent of `apt-get upgrade`

Simply running `apt-offline set /tmp/set.uris" implies both --update and --upgrade.

If your apt database is up-to-date from the last run, and you only want the list of packages to be upgraded, you should run: `apt-offline set /tmp/upgrade.uris --upgrade`. This will generate a list with all upgradable packages.

**Comment by rickysarraf** [comment]
https://framabin.org/?0e346d5271fc56ef#k/K9VBHHSrtYQhS/kUumKyZ+BvjfyiXjwU5xg1U0Xa8=

This link is only run with `apt-offline set update.uris --update`, which only updates your apt database.

**Comment by rickysarraf** [comment]
https://framabin.org/?31aa24aa8fa19154#WFBz3cvaUaxqf6zfbYLnOD3BfRf2PzfP4UEGS6N5K7M=

And as you can see, because you had used only the `--update` option, in the install target, only the apt database was updated.

When you update the apt database, it gets informed of new upgradable packages, that apt-offline has not downloaded yet.


In the apt-offline manpage, I've given an example sequence, which I think may be the cause of the confusion.

1st sequence should be:

`apt-offline set /tmp/update.uris --update`
`apt-offline get update.uris --bundle update.zip`
`apt-offline install update.zip`

The above sequence when completed will update your disconnected machine's APT database.

2nd sequence should be:

`apt-offline set /tmp/upgrade.uris --upgrade`
`apt-offline get upgrade.uris --bundle upgrade.zip`
`apt-offline install upgrade.zip`

This will get you all the packages available on the disconnected machine.


After 1st and 2nd, when you run `apt-get upgrade`, it should show that all packages are locally available and that nothing else needs to be downloaded.

**Comment by rickysarraf** [comment]
Note to self: If this sequence is confirmed to be simpler, then I need to update the manpage accordingly.

**Comment by sylvainpelissier** [comment]
Ok thank you for your answer. With the sequence:
```
apt-offline set /tmp/upgrade.uris
apt-offline get upgrade.uris --bundle upgrade.zip
apt-offline install upgrade.zip
````

And finally:
```
apt-get upgrade
````

It worked fine. Indeed I did not understand correctly the manpage

**Comment by rickysarraf** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

On Thu, 2017-01-05 at 03:14 -0800, Sylvain Pelissier wrote:
> And finally:
> apt-get upgrade
> It worked fine. Indeed I did not understand correctly the manpage

I'm hoping you to give me some input, preferably in plain english, on what to
improve in the wordings.

- -- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEQCVDstmIVAB/Yn02pjpYo/LhdWkFAlhuK58ACgkQpjpYo/Lh
dWnj7g//TwtuyQr8iwY40cZWRoeAl95NvflRSNaaunVNl9hIjkkt6mBOCQNq7yIm
bWAhZ/FBUFe5vjeGs4zVGhkfBCmDATNBCuuz+yYPxsb5J3F29ZRG/V9RHoi6g6KW
n1F5Hn15jM568pImC+Kx7xQ09jACzPLfQKyTGR15ARRTpUDRpjVYuSXjIO3ahmCl
3QR48KDLmsUDfp70A/LGU7OeinfXBn2lmOUNW1qCFfjsZ4YsruTQ90W5RNBcVTXi
9/igeQg6GU/0jP6dwJ4F6vvWOEJcJLy2wqWd7nJpUxRc8LfbMOPwM9zKVSYLMMG7
aF3D9eMzYqPrql7+glyKZJzC2EQsb6PSZgugePvzNqXDYDwpeK+nS+fgMyHan0KM
n38dYCXOwd1nKmkrYp3ntCuwR//pCmsn72Vl7LGP6KohAMysf5a37jQo8+2iffDY
Z+eaZe1WNK4AfD7ahHCVRe/io+rGR9q9vQ3LXnsAOMuKp21wTEoSzi1lFdZ5Y5wY
fPSZn2WjF1Qy91AZ5r9RH2an/KTUx/xjCHhHMJ9FmbBRbEn5DaGbAt3gi1+gmoF+
B2B1NF17boKPg5+yTPPTCgod2ua3iPHZdXKzP5c0UM0ocO3jdSQkEs/c9OWz0VVz
AvJT9B44Flx4oolLRZH4txDJHW9UjfWHJwDj2Wt9TFKNknw78QQ=
=wBmv
-----END PGP SIGNATURE-----

**Comment by rickysarraf** [comment]
@sylvainpelissier I've added the details about dependencies to INSTALL file, which is where usually such information is mentioned.

**Comment by sylvainpelissier** [comment]
Thank you.
